### ISSUE TITLE
Looping through annotations on a page

### ISSUE THREAD

**Issue by merch89** [rk700/PyMuPDF#222]
Hi, 

First-time poster on Git, so I apologize in advance if this not the right venue. 

I am trying to loop through all of the annotations on a given page and changing the color if the 1) the annotation exists and 2) it is a highlight annotation (i.e., annot.type[0] == 8).

```
page = doc[0]
annot = page.firstAnnot

if annot:
    if annot .type[0] == 8:
        annot.setColors({'stroke':[128,0,128]})
        annot.updateAppearance()
    new = annot.next
    if new:
        if new.type[0] == 8:
            new.setColors({'stroke':[128,0,128]})
            new.updateAppearance()
```

The above finds the first highlight annotation and changes the color, finds the next highlight and changes the color. However, the use case is to loop through all annotations on a page and all pages within a document dynamically. 

Is there a way to capture all of the annotations on a page? If only to get a number that I use a for/while loop on as a control mechanism?

Thanks,

merch89

**Comment by JorjMcKie** [comment]
Hi,
Answering via e-mail – my internet connection is horrible again.
This is the right venue – no worries 😊

Annotations and links are separate, forward connected chains of object – using their “next” property. You can easily iterate through annots like so:

annot = page.firstAnnot
while annot:
    if annot.type[0] = correct:
        # do your stuff with the annot
    annot = annot.next

The loop automatically stopsat the last annotation, because is “next” property is None.

The colors in PDF are represented by sequences of float values between 0 and 1 (including both). So instead [128, 0, 128], you must use something like [128./255, 0, 128./255].
Instead of “annot.updateAppearance” please use “annot.update” (at least starting with the coming v1.14.0)

To loop through all annots of all pages do this:

for page in doc:
    annot = page.firstAnnot
    while annot:
        # the above loop

After being done, do not forget to save the PDF – either to a new file or incrementally.


Gesendet von Mail<https://go.microsoft.com/fwlink/?LinkId=550986> für Windows 10

________________________________
Von: merch89 <notifications@github.com>
Gesendet: Friday, November 16, 2018 11:18:25 AM
An: rk700/PyMuPDF
Cc: Subscribed
Betreff: [rk700/PyMuPDF] Looping through annotations on a page (#222)


Hi,

First-time poster on Git, so I apologize in advance if this not the right venue.

I am trying to loop through all of the annotations on a given page and changing the color if the 1) the annotation exists and 2) it is a highlight annotation (i.e., annot.type[0] == 8).

page = doc[0] annot = page.firstAnnot if annot: if annot .type[0] == 8: annot.setColors({'stroke':[128,0,128]}) annot.updateAppearance() new = annot.next if new: if new.type[0] == 8: new.setColors({'stroke':[128,0,128]}) new.updateAppearance()

The above finds the first highlight annotation and changes the color, finds the next highlight and changes the color. However, the use case is to loop through all annotations on a page and all pages within a document dynamically.

Is there a way to capture all of the annotations on a page? If only to get a number that I use a for/while loop on as a control mechanism?

Thanks,

merch89

**Comment by merch89** [comment]
Hi, Jorj,

Thank you for the quick reply, I really appreciate it! Ironically, I was reading your article at http://code.activestate.com/recipes/580732-access-pdf-annotations/, when I got your response.

I made the changes to the color reference as you suggested.

As for the updateAppreance() vs. update(), I will keep an eye out for the change we update the PyMuPDF module. 

Thanks again for your help!

Best,

merch89
